lib/pure/unicode.nim

```diff
@@ -1423,6 +1413,47 @@ proc isSpace*(s: string): bool {.noSideEffect, procvar,
   ## Returns true iff `s` contains all whitespace unicode characters.
   runeCheck(s, isWhiteSpace)
 
+template runeCaseCheck(s, runeProc) =
+  ## Common code for rune.isLower and rune.isUpper.
+  if len(s) == 0: return false
+
+  var
+    i = 0
+    rune: Rune
+    hasAtleastOneAlphaRune = false
+    doCaseCheck: bool
+
+  while i < len(s):
+    fastRuneAt(s, i, rune, doInc=true)
+    doCaseCheck = isAlpha(rune)
+    if not hasAtleastOneAlphaRune:
+      hasAtleastOneAlphaRune = doCaseCheck
+    if doCaseCheck and (not runeProc(rune)): return false
+  if not hasAtleastOneAlphaRune:
```

Comment:
I could have just returned `hasAtleastOneAlphaRune` here. Will update the PR tomorrow.